### ISSUE TITLE
hx-target previous behavior with nested matches

### ISSUE THREAD

**Issue by Trildar** [bigskysoftware/htmx#1780]
`previous` currently select the most deeply nested element matching the selector preceding the element with `hx-target`. I was not expecting this behavior, particularly in the case when there are sibling elements matching the selector. The reference for `hx-target` even mentions sibling elements: "(e.g previous .error will target the closest previous sibling with error class)".

```HTML
<div>
    <div class="target" id="1">
        <div class="target" id="2"></div>
    </div>
    <button hx-target="previous .target">Hello</button>
</div>
```

For example, in the above, the target of the button will be the inner div with id="2", instead of the outer div with id="1".

```HTML
<div class="target" id="1">
    <div class="target" id="2"></div>
</div>
<div>
    <button hx-target="previous .target">Hello</button>
</div>
```

Similarly, in this case, the target will again be the inner div with id="2", though I'm not sure selecting the outer div would be more commonly expected in this case. Though it would be more consistent with the behavior of selecting siblings first, as it would be something like "traversing the DOM tree backwards starting from `this`, without stepping into any deeper nesting".

That said though, I'm not sure if the current behavior is something that should actually be changed, or if it just needs to be documented better.

---

EDIT: Right after posting, I realized that the way `previous` behaves is, in a sense, problematic due to how it contrasts with the behavior of `next` which instead has preference for less deeply nested nodes.

```HTML
<div>
    <button hx-target="next .target">Hello</button>
    <div class="target" id="1">
        <div class="target" id="2"></div>
    </div>
</div>
```
```HTML
<div>
    <button hx-target="next .target">Hello</button>
</div>
<div class="target" id="1">
    <div class="target" id="2"></div>
</div>
```

In these two examples, the divs with id="1" are the targets.

**Comment by Telroshan** [comment]
Hey, looking at [the code](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L609), htmx relies on the [Node.DOCUMENT_POSITION_PRECEDING](https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition#return_value) and [Node.DOCUMENT_POSITION_FOLLOWING](https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition#return_value) constants

`DOCUMENT_POSITION_PRECEDING`:
> as an ancestor or previous sibling or a descendant of a previous sibling or previous sibling of an ancestor

`DOCUMENT_POSITION_FOLLOWING`:
> as a descendant or following sibling or a descendant of a following sibling or following sibling of an ancestor

So it seems the behaviour you're observing is working as intended
I'm not sure how to feel about this, I never got in this situation of nested identically-classed elements, not sure what would be "the right way" to handle that

Intuitively, I would have also expected `next` and `previous` to only look at siblings of the same parents, and not descendants of the siblings. Though, I didn't write this feature and I'm probably missing some usecase/context as per its implementation

What is also sure btw, is that modifying the current behaviour would be a breaking change, so even if we were to change that, it would require some config param, or an entirely new keyword to handle this probably. Keep that in mind if you'd like to send  PR to address that issue!

May I invoke you @1cg on this matter ?

**Comment by Telroshan** [comment]
I think one possible approach to implement that, would be to loop on the source element's [`previousElementSibling`](https://developer.mozilla.org/en-US/docs/Web/API/Element/previousElementSibling) or [nextElementSibling](https://developer.mozilla.org/en-US/docs/Web/API/Element/nextElementSibling), and use [element.matches](https://developer.mozilla.org/fr/docs/Web/API/Element/matches) with the given selector. Loop until finding a sibling that matches the selector, or until null (i.e no more previous/next sibling), then return null

**Comment by Trildar** [comment]
Ah, right. Missed `Element.matches` in my brief search before. Anyway, I'll wait on discussion on what, if anything, should change.

I suppose I could make a PR for the website docs. Not sure how to describe it concisely though. Something like "Note that if the selector matches both an element and a descendant of that element, the descendant will be the target", added to the description for `previous`?

**Comment by Telroshan** [comment]
Maybe illustrating it with a short example would help making it clear for readers ? Like:

Be aware than in this situation
```html
<div id="parent" class="some-class">
    <div id="child" class="some-class"></div>
</div>
<button hx-target="previous .some-class" hx-get="/url">Click me</button>
```
Clicking on the button would target the `child` element and not the `parent` one, as it is closer in a flattened hierarchy:
- parent
- child
- button

**Comment by Trildar** [comment]
Yeah, example would be the clearest, but I was just thinking a concise sentence would be easier to insert on either the description of `previous` or in the Notes section. I guess put it before Notes with a header like "Nested Matches and previous"?

**Comment by Telroshan** [comment]
Just remembered we talked a while ago about those extended selectors' documentation in #1601
As previous/next are usable within several hx attributes

If and when that happens, going for the not-concise but clearest way should be fine I guess, at it'd be a page dedicated to explaining those special selectors

**Comment by 1cg** [comment]
no longer relevant
